### PR TITLE
Fix memory leaks reported by ASAN (with suppressions turned off) part 2

### LINKED ISSUE #10484: Memory leak in DDL changing space
Found by ASAN.

Reproducer:

```lua
local fiber = require('fiber')
box.cfg{}
box.execute('CREATE TABLE t1 (a INT PRIMARY KEY);')
box.execute('CREATE VIEW v AS SELECT * FROM t1;')
while true do
    for i = 1,1000 do
        box.execute('ALTER TABLE v ADD c INT;')
    end
    collectgarbage()
    fiber.yield()
end
```

The script memory usage uniformly grows by 40Mb/s (debug build).

Introduced in `3.0.0-alpha2-43-g96a5224a40`.

### LINKED ISSUE #10485: Memory leak in DDL changing user
Found by ASAN.

Reproducer:

```lua
local fiber = require('fiber')
box.cfg{}
box.schema.user.create('grantor')
box.schema.user.grant('grantor', 'read, write, execute, alter, drop', 'universe')
box.session.su('grantor')
while true do
    for i = 1,1000 do
        pcall(box.schema.user.create, 'grantee')
    end
    collectgarbage()
    fiber.yield()
end
```
The script memory usage uniformly grows by 2Mb/s (debug build).

Introduced in `1.7.7-141-g5b32bb7ff5`.

### LINKED ISSUE #10487: Memory leak in lua iconv
Found by ASAN.

Reproducer:
```lua
local iconv = require('iconv')
local fiber = require('fiber')
while true do
    for i = 1,1000 do
        iconv.new('ASCII', 'UTF-8')
    end
    collectgarbage();
    fiber.yield()
end
```

The script is killed by OOM killer after dozen of seconds after start.
```
авг 28 16:00:13 mynote kernel: Out of memory: Killed process 551488 (tarantool) total-vm:84763856kB, anon-rss:11521760kB, file-rss:2280kB, shmem-rss:0kB, UID:1000 pgtables:165936kB oom_score_adj:0
```

Introduced in `1.7.4-262-gaff6235cf0`.

### LINKED ISSUE #10489: Memory leak in vinyl on dump/compaction failure
Reported by ASAN for `vinyl/recovery_quota.test.lua` test.

```
2024-07-05 13:30:34.605 [478603] main/103/on_shutdown vy_scheduler.c:1668 E> 512/0: failed to compact range (-inf..inf)

=================================================================
==478603==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 4 byte(s) in 1 object(s) allocated from:
    #0 0x5e4ebafcae09 in malloc (/home/shiny/dev/tarantool/build-asan-debug/src/tarantool+0x1244e09) (BuildId: 20c5933d67a3831c4f43f6860379d58d35b81974)
    #1 0x5e4ebb3f9b69 in vy_key_dup /home/shiny/dev/tarantool/src/box/vy_stmt.c:308:14
    #2 0x5e4ebb49b615 in vy_page_info_create /home/shiny/dev/tarantool/src/box/vy_run.c:257:23
    #3 0x5e4ebb48f59f in vy_run_writer_start_page /home/shiny/dev/tarantool/src/box/vy_run.c:2196:6
    #4 0x5e4ebb48c6b6 in vy_run_writer_append_stmt /home/shiny/dev/tarantool/src/box/vy_run.c:2287:6
    #5 0x5e4ebb72877f in vy_task_write_run /home/shiny/dev/tarantool/src/box/vy_scheduler.c:1132:8
    #6 0x5e4ebb73305e in vy_task_compaction_execute /home/shiny/dev/tarantool/src/box/vy_scheduler.c:1485:9
    #7 0x5e4ebb73e152 in vy_task_f /home/shiny/dev/tarantool/src/box/vy_scheduler.c:1795:6
    #8 0x5e4ebb01e0b1 in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*) /home/shiny/dev/tarantool/src/lib/core/fiber.h:1331:10
    #9 0x5e4ebc389ee0 in fiber_loop /home/shiny/dev/tarantool/src/lib/core/fiber.c:1182:18
    #10 0x5e4ebd3e9595 in coro_init /home/shiny/dev/tarantool/third_party/coro/coro.c:108:3

SUMMARY: AddressSanitizer: 4 byte(s) leaked in 1 allocation(s).
```

Can be reproduced with #10207 without `vinyl: fix memory leak on dump/compaction failure` commit.

Introduced in `1.7.2-405-g9b99d64b0`.

### LINKED ISSUE #10490: Memory leak when replica notice it is removed from cluster
Reported by ASAN for `replication-luatest/gh_10088_apply_deletion_of_replica_from_cluster_on_deleted_replica_test.lua`  test.

```
=================================================================
==83862==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 872 byte(s) in 1 object(s) allocated from:
    #0 0x55ba0ce6bcae in malloc (/__w/tarantool/tarantool/src/tarantool+0x11dbcae) (BuildId: 2d31b86dbd588e22ddde87ef7df6f8c4d9c094c9)
    #1 0x55ba0e2b61cc in Exception::operator new(unsigned long) /__w/tarantool/tarantool/src/lib/core/exception.cc:90:9
    #2 0x55ba0e2bd1d2 in BuildSocketError /__w/tarantool/tarantool/src/lib/core/exception.cc:397:19
    #3 0x55ba0e26befa in coio_readn_ahead_timeout /__w/tarantool/tarantool/src/lib/core/coio.c:357:3
    #4 0x55ba0cf51644 in coio_breadn_timeout(iostream*, ibuf*, unsigned long, double) /__w/tarantool/tarantool/src/lib/core/coio_buf.h:98:14
    #5 0x55ba0cf50675 in coio_read_xrow_timeout_xc /__w/tarantool/tarantool/src/box/xrow_io.cc:74:3
    #6 0x55ba0d99b2ad in applier_read_tx_row(applier*, applier_read_ctx const*, double) /__w/tarantool/tarantool/src/box/applier.cc:1105:2
    #7 0x55ba0d998cf8 in applier_read_tx(applier*, stailq*, applier_read_ctx const*, double) /__w/tarantool/tarantool/src/box/applier.cc:1202:4
    #8 0x55ba0d9b80d4 in applier_thread_reader_f(__va_list_tag*) /__w/tarantool/tarantool/src/box/applier.cc:2070:4
    #9 0x55ba0ceb1468 in fiber_cxx_invoke(int (*)(__va_list_tag*), __va_list_tag*) /__w/tarantool/tarantool/src/lib/core/fiber.h:1324:10
    #10 0x55ba0e1e7367 in fiber_loop /__w/tarantool/tarantool/src/lib/core/fiber.c:1162:18
    #11 0x55ba0f21b175 in coro_init /__w/tarantool/tarantool/third_party/coro/coro.c:108:3

SUMMARY: AddressSanitizer: 872 byte(s) leaked in 1 allocation(s).
```

Can be reproduced with https://github.com/tarantool/tarantool/pull/10207.

Introduced in `3.2.0-entrypoint-173-g25aac111ce`.

### PR DESCRIPTION
**Pull request by nshy**
Part of https://github.com/tarantool/tarantool/issues/10211

Factored out from all-in-one PR https://github.com/tarantool/tarantool/pull/10207

Closes #10484
Closes #10485 
Closes #10487 
Closes #10489 
Closes #10490

### CODE REVIEW

File: src/box/space_def.c [hunk 1]
```diff
@@ -223,6 +223,8 @@ space_def_new_ephemeral(uint32_t exact_field_count, struct field_def *fields)
 void
 space_def_delete(struct space_def *def)
 {
+	if (def == NULL)
+		return;
```
Comment (locker) [inline src/box/space_def.c:227]: I don't like that this function handles `def = NULL` while, for example, `index_def_delete` or `key_def_delete`. Please move this check to where this function is called.

File: changelogs/unreleased/gh-10484-fix-space-DDL-memleak.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
+
+- Fixed a memory leak on space DDL under certain conditions (gh-10484).
```
Comment (locker) [inline changelogs/unreleased/gh-10484-fix-space-DDL-memleak.md:3]: Bad change log. Please be more specific.

File: changelogs/unreleased/gh-10485-fix-user-DDL-memleak.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+## bugfix/box
+
+- Fixed a memory leak on user DDL when access is denied (gh-10485).
```
Comment (locker) [inline changelogs/unreleased/gh-10485-fix-user-DDL-memleak.md:3]: Bad wording: DDL is "data definition language" so the sentence doesn't make any sense. Should be something like "Fixed a memory leak on an attempt to create or modify a user denied due to access restrictions". The doc team may suggest a better variant.

File: changelogs/unreleased/gh-10489-fix-vinyl-on-dump-failure-memleak.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+## bugfix/vinyl
+
+- Fixed a memory leak in vinyl on dump/compaction failure (gh-10489).
```
Comment (locker) [inline changelogs/unreleased/gh-10489-fix-vinyl-on-dump-failure-memleak.md:3]: "in vinyl" is redundant here - this changelog entry is for Vinyl.

File: changelogs/unreleased/gh-10487-fix-iconv-memleak.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+## bugfix/lua
+
+- Fixed a memory leak in iconv (gh-10487).
```
Comment (locker) [inline changelogs/unreleased/gh-10487-fix-iconv-memleak.md:3]: in `iconv.new()`

**Review by p7nov — APPROVED** [review]

File: changelogs/unreleased/gh-10490-fix-replica-deleted-memleak.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+## bugfix/replication
+
+- Fixed a memory leak when replica noticed it is deleted from cluster
```
Comment (locker) [inline changelogs/unreleased/gh-10490-fix-replica-deleted-memleak.md:3]: Fixed a memory leak on a replication error.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/space_def.c (hunk 1 above)
Comment (nshy) [reply src/box/space_def.c:227]: Fixed.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index b413fe42e3..a28d750ee7 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -2116,8 +2116,10 @@ on_replace_dd_space(struct trigger * /* trigger */, void *even>
                        ER_CREATE_SPACE : ER_ALTER_SPACE;
                def = space_def_new_from_tuple(new_tuple, errcode, region);
        }
-       auto def_guard =
-               make_scoped_guard([=] { space_def_delete(def); });
+       auto def_guard = make_scoped_guard([=] {
+               if (def != NULL)
+                       space_def_delete(def);
+       });
        if (filter_temporary_ddl_stmt(txn, old_space != NULL ?
                                      old_space->def : def) != 0)
                return -1;
diff --git a/src/box/space_def.c b/src/box/space_def.c
index 945d4c31c3..3dd9da133b 100644
--- a/src/box/space_def.c
+++ b/src/box/space_def.c
@@ -223,8 +223,6 @@ space_def_new_ephemeral(uint32_t exact_field_count, struct field_>
 void
 space_def_delete(struct space_def *def)
 {
-       if (def == NULL)
-               return;
        field_def_array_delete(def->fields, def->field_count);
        tuple_dictionary_unref(def->dict);
        free(def->opts.sql);
```

</details>

File: changelogs/unreleased/gh-10484-fix-space-DDL-memleak.md (hunk 2 above)
Comment (nshy) [reply changelogs/unreleased/gh-10484-fix-space-DDL-memleak.md:3]: Changed to:

```
## bugfix/sql

- Fixed a memory leak on attempt to alter table view (gh-10484).
```

File: changelogs/unreleased/gh-10485-fix-user-DDL-memleak.md (hunk 3 above)
Comment (nshy) [reply changelogs/unreleased/gh-10485-fix-user-DDL-memleak.md:3]: Fixed as suggested.

File: changelogs/unreleased/gh-10489-fix-vinyl-on-dump-failure-memleak.md (hunk 4 above)
Comment (nshy) [reply changelogs/unreleased/gh-10489-fix-vinyl-on-dump-failure-memleak.md:3]: Fixed as suggested.

File: changelogs/unreleased/gh-10487-fix-iconv-memleak.md (hunk 5 above)
Comment (nshy) [reply changelogs/unreleased/gh-10487-fix-iconv-memleak.md:3]: Fixed as suggested.

File: src/box/applier.cc [hunk 7]
```diff
@@ -2296,6 +2296,7 @@ applier_thread_data_destroy(struct applier *applier)
 		say_warn("applier data destruction is continued");
 	});
 
+	diag_clear(&applier->thread.exit_msg.diag);
```
Comment (locker) [inline src/box/applier.cc:2299]: Let's move it to `applier_thread_detach_applier` for the sake of symmetry.

File: changelogs/unreleased/gh-10490-fix-replica-deleted-memleak.md (hunk 6 above)
Comment (nshy) [reply changelogs/unreleased/gh-10490-fix-replica-deleted-memleak.md:3]: Fixed as suggested.

File: src/box/applier.cc (hunk 7 above)
Comment (nshy) [reply src/box/applier.cc:2299]: Fixed.

```diff
--- a/src/box/applier.cc
+++ b/src/box/applier.cc
@@ -2269,6 +2269,7 @@ applier_thread_detach_applier(struct cbus_call_msg *base)
        lsregion_destroy(&applier->thread.lsr);
        fiber_cond_destroy(&applier->thread.writer_cond);
        ibuf_destroy(&applier->thread.ibuf);
+       diag_clear(&applier->thread.exit_msg.diag);

        return 0;
 }
@@ -2296,7 +2297,6 @@ applier_thread_data_destroy(struct applier *applier)
                say_warn("applier data destruction is continued");
        });

-       diag_clear(&applier->thread.exit_msg.diag);
        fiber_cond_destroy(&applier->msg_cond);
 }
```

File: changelogs/unreleased/gh-10484-fix-space-DDL-memleak.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+## bugfix/sql
+
+- Fixed a memory leak on attempt to alter table view (gh-10484).
```
[the remark this answers is not in the document]
Comment (nshy) [reply changelogs/unreleased/gh-10484-fix-space-DDL-memleak.md:3]: Fixed as suggested.

File: changelogs/unreleased/gh-10487-fix-iconv-memleak.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+## bugfix/lua
+
+- Fixed a memory leak in iconv.new() (gh-10487).
```
[the remark this answers is not in the document]
Comment (nshy) [reply changelogs/unreleased/gh-10487-fix-iconv-memleak.md:3]: Fixed.

**Comment by locker** [comment]
Cherry-picked to 2.11 and 3.2.

**Pull request merged**